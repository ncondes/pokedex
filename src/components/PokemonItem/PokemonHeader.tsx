'use client'

import { FC } from 'react'
import { capitalize } from '@/utils'
import styles from './PokemonItem.module.scss'

interface Props {
   id: number
   name: string
}

export const PokemonHeader: FC<Props> = ({ id, name }) => {
   const isFavorite = true
   const isLoading = false

   const handleFavoriteButton = () => {
      // TODO: handleFavoriteButton
      console.log('Favorite')
   }

   return (
      <div className={styles.pokemon_header}>
         <h2>{capitalize(name)}</h2>
         <button onClick={handleFavoriteButton} disabled={isLoading}>
            {isLoading ? (
               <i className="fa-solid fa-spinner fa-spin"></i>
            ) : (
               <i
                  className={`${
                     isFavorite ? 'fa-solid' : 'fa-regular'
                  } fa-heart fa-xl`}
               ></i>
            )}
         </button>
      </div>
   )
}
