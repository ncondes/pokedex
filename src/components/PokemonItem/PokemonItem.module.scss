@use '@/scss/colors' as *;

.pokemon_card {
   display: grid;
   grid-template-rows: repeat(2, 1fr);
   width: 100%;
   height: 30rem;
   border-radius: 1rem;
   box-shadow: 0px 10px 15px -3px rgba(0, 0, 0, 0.1);
}

.pokemon_header {
   display: flex;
   justify-content: space-between;
   align-items: center;
   margin-top: 1.5rem;
   padding: 0 2rem;

   h2 {
      color: $white-cream-color;
   }

   button {
      background-color: transparent;
      z-index: 1;
   }

   i {
      color: $white-cream-color;
   }
}

.pokemon_chips {
   margin-top: 0.3rem;
   padding: 0 2rem;
}

.pokemon_chip {
   display: inline-block;
   background-color: rgba($white-cream-color, 0.3);
   margin-right: 0.5rem;
   padding: 0.3rem 1rem;
   border-radius: 1rem;
   font-size: 0.7rem;
   font-weight: 700;
   color: $white-cream-color;
}

.white_transparent_icon {
   > img {
      filter: invert(100%) opacity(0.25);
   }
}

.pokemon_bg_image_container {
   display: flex;
   justify-content: center;
   align-items: center;
   position: relative;
}

.pokeball_icon {
   position: absolute;
   right: 0rem;
   bottom: 2rem;
   @extend .white_transparent_icon;
}

.points_icon {
   position: absolute;
   left: 2rem;
   top: 1.2rem;
   @extend .white_transparent_icon;
}

.lines_icon {
   position: absolute;
   right: 0.5rem;
   top: -4.5rem;
   z-index: 0;
   @extend .white_transparent_icon;
}

.pokemon_image {
   z-index: 1;
}

.pokemon_stats {
   margin-top: -1.5rem;
   border-radius: 1rem;
   padding: 2rem;

   p {
      font-weight: 700;
      font-size: 0.8rem;
      margin-bottom: 1rem;
   }

   > div {
      display: grid;
      place-items: center;
      gap: 0.5rem;
      width: 100%;
      grid-template-rows: repeat(6, 1fr);
   }
}

.pokemon_stat {
   display: grid;
   width: 100%;
   grid-template-columns: 2fr 1fr 4fr;

   span {
      font-weight: 300;
      font-size: 0.8rem;
   }

   b {
      font-weight: 700;
      font-size: 0.8rem;
   }
}

.pokemon_statbar_container {
   display: flex;
   align-items: center;
   position: relative;
}

.pokemon_statbar_bar {
   background-color: #e0e0e0;
   border-radius: 0.5rem;
   height: 0.3rem;
   width: 100%;
   position: absolute;

   > div {
      height: 100%;
      border-radius: 0.5rem;
   }
}
