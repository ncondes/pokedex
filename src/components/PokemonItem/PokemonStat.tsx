import { FC, useContext } from 'react'
import { capitalize } from '@/utils'
import { ThemeContext } from '@/context/theme'
import styles from './PokemonItem.module.scss'

interface Props {
   stat: string
   value: number
}

export const PokemonStat: FC<Props> = ({ stat, value }) => {
   const { useTheme } = useContext(ThemeContext)

   let background: string = 'high'

   if (value < 30) {
      background = '#d85248'
   } else if (value >= 30 && value <= 50) {
      background = '#dfd158'
   } else if (value > 50) {
      background = '#88c45f'
   }

   const formatStatName = (name: string) => {
      if (name === 'special-attack') return 'Sp. Atk'
      if (name === 'special-defense') return 'Sp. Def'
      return capitalize(name)
   }

   return (
      <div className={`${styles.pokemon_stat} ${useTheme('font')}`}>
         <span>{formatStatName(stat)}</span>
         <b>{value}</b>
         {/* stat bar */}
         <div className={styles.pokemon_statbar_container}>
            <div className={styles.pokemon_statbar_bar}>
               <div
                  style={{ width: `${Math.min(value, 100)}%`, background }}
               ></div>
            </div>
         </div>
      </div>
   )
}
