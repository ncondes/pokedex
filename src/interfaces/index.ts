export * from './data'
export * from './pokemon'
