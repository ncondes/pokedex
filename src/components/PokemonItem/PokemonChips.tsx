import { Type } from '@/interfaces'
import { FC } from 'react'
import { capitalize } from '@/utils'
import styles from './PokemonItem.module.scss'

interface Props {
   types: Type[]
}

export const PokemonChips: FC<Props> = ({ types }) => {
   return (
      <div className={styles.pokemon_chips}>
         {types.map(({ type }) => (
            <div key={type.name} className={styles.pokemon_chip}>
               {capitalize(type.name)}
            </div>
         ))}
      </div>
   )
}
