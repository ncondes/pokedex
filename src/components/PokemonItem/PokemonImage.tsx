import Image from 'next/image'
import { Sprites } from '@/interfaces'
import { FC } from 'react'
import styles from './PokemonItem.module.scss'

interface Props {
   sprites: Sprites
}

export const PokemonImage: FC<Props> = ({ sprites }) => {
   return (
      <div className={styles.pokemon_bg_image_container}>
         {/* background icons */}
         <div className={styles.pokeball_icon}>
            <Image
               src={'/images/pokeball-icon.png'}
               alt="pokeball icon"
               width={150}
               height={150}
            />
         </div>
         <div className={styles.points_icon}>
            <Image
               src={'/images/points-icon.png'}
               alt="points icon"
               width={50}
               height={50}
            />
         </div>
         <div className={styles.lines_icon}>
            <Image
               src={'/images/lines-icon.png'}
               alt="lines icon"
               width={50}
               height={50}
            />
         </div>
         {/* pokemon image */}
         <div className={styles.pokemon_image}>
            <Image
               src={sprites.other?.home?.front_default || sprites.front_default}
               alt="pokemon image"
               width={200}
               height={200}
               priority
            />
         </div>
      </div>
   )
}
