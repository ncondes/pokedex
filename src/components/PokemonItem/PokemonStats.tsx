'use client'

import { FC, useContext } from 'react'
import { ThemeContext } from '@/context/theme'
import { Stat } from '@/interfaces'
import styles from './PokemonItem.module.scss'
import { PokemonStat } from './PokemonStat'

interface Props {
   stats: Stat[]
}

export const PokemonStats: FC<Props> = ({ stats }) => {
   const { useTheme } = useContext(ThemeContext)

   return (
      <div className={`${styles.pokemon_stats} ${useTheme('card-bg')}`}>
         <p className={useTheme('font')}>Base Stats</p>
         <div className={styles.base_stats}>
            {stats.map(({ stat, base_stat }) => (
               <PokemonStat
                  key={stat.name}
                  stat={stat.name}
                  value={base_stat}
               />
            ))}
         </div>
      </div>
   )
}
