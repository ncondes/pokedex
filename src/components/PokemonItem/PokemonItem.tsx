import { Pokemon } from '@/interfaces'
import { FC } from 'react'
import { PokemonHeader } from './PokemonHeader'
import { PokemonChips } from './PokemonChips'
import { PokemonImage } from './PokemonImage'
import styles from './PokemonItem.module.scss'
import { PokemonStats } from './PokemonStats'

interface Props {
   pokemon: Pokemon
}

export const PokemonItem: FC<Props> = ({ pokemon }) => {
   const type = `pokemon-type-${pokemon.types?.[0].type.name || 'default'}`

   return (
      <article className={`${styles.pokemon_card} ${type}`}>
         {/* pokemon image */}
         <section>
            <PokemonHeader id={pokemon.id} name={pokemon.name} />
            <PokemonChips types={pokemon.types} />
            <PokemonImage sprites={pokemon.sprites} />
         </section>
         {/* pokemon description */}
         <PokemonStats stats={pokemon.stats} />
      </article>
   )
}
