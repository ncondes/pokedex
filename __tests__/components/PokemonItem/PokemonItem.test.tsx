import React from 'react'
import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import { PokemonItem } from '@/components/PokemonItem/PokemonItem'
import { Pokemon } from '@/interfaces'
import { ThemeProvider } from '@/context/theme'

const pokemon = {
   id: 1,
   name: 'Pikachu',
   types: [{ type: { name: 'electric' } }],
   sprites: {
      front_default:
         'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/25.png',
   },
   stats: [
      { stat: { name: 'hp' }, base_stat: 35 },
      { stat: { name: 'attack' }, base_stat: 55 },
      { stat: { name: 'defense' }, base_stat: 40 },
   ],
} as Pokemon

describe('PokemonItem', () => {
   it('renders the Pokemon details correctly', () => {
      const { getByText, getByAltText } = render(
         <ThemeProvider>
            <PokemonItem pokemon={pokemon} />
         </ThemeProvider>
      )

      expect(getByText('Pikachu')).toBeInTheDocument()
      expect(getByAltText('pokemon image')).toBeInTheDocument()
      expect(getByText('Hp')).toBeInTheDocument()
      expect(getByText('35')).toBeInTheDocument()
      expect(getByText('Attack')).toBeInTheDocument()
      expect(getByText('55')).toBeInTheDocument()
      expect(getByText('Defense')).toBeInTheDocument()
      expect(getByText('40')).toBeInTheDocument()
   })

   it('renders the correct CSS class based on the Pokemon type', () => {
      const { container } = render(
         <ThemeProvider>
            <PokemonItem pokemon={pokemon} />
         </ThemeProvider>
      )

      expect(container.firstChild).toHaveClass('pokemon-type-electric')
   })

   it('should match the snapshot', () => {
      const { container } = render(
         <ThemeProvider>
            <PokemonItem pokemon={pokemon} />
         </ThemeProvider>
      )

      expect(container.firstChild).toMatchSnapshot()
   })
})
