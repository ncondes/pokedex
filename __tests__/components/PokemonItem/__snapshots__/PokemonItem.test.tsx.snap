// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonItem should match the snapshot 1`] = `
<article
  class="pokemon_card pokemon-type-electric"
>
  <section>
    <div
      class="pokemon_header"
    >
      <h2>
        Pikachu
      </h2>
      <button>
        <i
          class="fa-regular fa-heart fa-xl"
        />
      </button>
    </div>
    <div
      class="pokemon_chips"
    >
      <div
        class="pokemon_chip"
      >
        Electric
      </div>
    </div>
    <div
      class="pokemon_bg_image_container"
    >
      <div
        class="pokeball_icon"
      >
        <img
          alt="pokeball icon"
          data-nimg="1"
          decoding="async"
          height="150"
          loading="lazy"
          src="/_next/image?url=%2Fimages%2Fpokeball-icon.png&w=384&q=75"
          srcset="/_next/image?url=%2Fimages%2Fpokeball-icon.png&w=256&q=75 1x, /_next/image?url=%2Fimages%2Fpokeball-icon.png&w=384&q=75 2x"
          style="color: transparent;"
          width="150"
        />
      </div>
      <div
        class="points_icon"
      >
        <img
          alt="points icon"
          data-nimg="1"
          decoding="async"
          height="50"
          loading="lazy"
          src="/_next/image?url=%2Fimages%2Fpoints-icon.png&w=128&q=75"
          srcset="/_next/image?url=%2Fimages%2Fpoints-icon.png&w=64&q=75 1x, /_next/image?url=%2Fimages%2Fpoints-icon.png&w=128&q=75 2x"
          style="color: transparent;"
          width="50"
        />
      </div>
      <div
        class="lines_icon"
      >
        <img
          alt="lines icon"
          data-nimg="1"
          decoding="async"
          height="50"
          loading="lazy"
          src="/_next/image?url=%2Fimages%2Flines-icon.png&w=128&q=75"
          srcset="/_next/image?url=%2Fimages%2Flines-icon.png&w=64&q=75 1x, /_next/image?url=%2Fimages%2Flines-icon.png&w=128&q=75 2x"
          style="color: transparent;"
          width="50"
        />
      </div>
      <div
        class="pokemon_image"
      >
        <img
          alt="pokemon image"
          data-nimg="1"
          decoding="async"
          fetchpriority="high"
          height="200"
          src="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F25.png&w=640&q=75"
          srcset="/_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F25.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fraw.githubusercontent.com%2FPokeAPI%2Fsprites%2Fmaster%2Fsprites%2Fpokemon%2F25.png&w=640&q=75 2x"
          style="color: transparent;"
          width="200"
        />
      </div>
    </div>
  </section>
  <div
    class="pokemon_stats light-card-bg"
  >
    <p
      class="light-font"
    >
      Base Stats
    </p>
    <div
      class="base_stats"
    >
      <div
        class="pokemon_stat light-font"
      >
        <span>
          Hp
        </span>
        <b>
          35
        </b>
        <div
          class="pokemon_statbar_container"
        >
          <div
            class="pokemon_statbar_bar"
          >
            <div
              style="width: 35%; background: rgb(223, 209, 88);"
            />
          </div>
        </div>
      </div>
      <div
        class="pokemon_stat light-font"
      >
        <span>
          Attack
        </span>
        <b>
          55
        </b>
        <div
          class="pokemon_statbar_container"
        >
          <div
            class="pokemon_statbar_bar"
          >
            <div
              style="width: 55%; background: rgb(136, 196, 95);"
            />
          </div>
        </div>
      </div>
      <div
        class="pokemon_stat light-font"
      >
        <span>
          Defense
        </span>
        <b>
          40
        </b>
        <div
          class="pokemon_statbar_container"
        >
          <div
            class="pokemon_statbar_bar"
          >
            <div
              style="width: 40%; background: rgb(223, 209, 88);"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</article>
`;
